csrc/optimization/mark_aliases_prepare.cpp

```diff
@@ -0,0 +1,114 @@
+
+namespace nvfuser::optimization {
+
+void MarkAliasesPreparePass::runPass(Fusion* fusion) {
+  const AliasAnalysisResult analysis =
+      findAliases(fusion, /*can_override_empty_allocation_domain=*/true);
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    debug() << "Fusion before MarkAliasesPreparePass:" << std::endl;
+    fusion->printMath();
+    debug() << "Alias analysis result:" << std::endl;
+    debug() << analysis.toString(/*indent_size=*/1) << std::endl;
+  }
+
+  // Fusion outputs that are (1) aliased by others and (2) not aliases
+  // themselves. Code will later add `segment_set` before them so aliases are
+  // separated from non-aliases and more likely to be accepted by the no-op
+  // scheduler.
+  std::vector<TensorView*> aliased_outs;
+  aliased_outs.reserve(fusion->outputs().size());
+
+  for (TensorView* out :
+       ir_utils::filterByType<TensorView>(fusion->outputs())) {
+    TensorView* aliased_io = analysis.getNearestAliasedIo(out);
+    if (aliased_io == nullptr) {
+      continue;
+    }
+
+    if (aliased_io->isFusionOutput() &&
+        analysis.getNearestAliasedIo(aliased_io) == nullptr) {
+      aliased_outs.push_back(aliased_io);
+    }
+
+    // We already checked it's compatible; no need to change.
+    if (out->hasAllocation()) {
+      continue;
+    }
+
+    // A scalar `out` triggers a corner case that crashes
+    // `validateDomainEquivalence`.
+    if (out->isZeroDim()) {
+      continue;
+    }
+
+    const optimization::Layout preferred_layout = analysis.preferredLayout(out);
+    out->setAllocationDomain(
+        preferred_layout.allocation_domain, preferred_layout.contiguity);
+    if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+      debug() << "Set the layout of " << ir_utils::varName(out) << " to "
+              << preferred_layout.toString() << std::endl;
+    }
+  }
+
+  for (TensorView* aliased_out : aliased_outs) {
+    // Rarely, if `aliased_out` is already defined by `segment_set`, don't
+    // create another `segment_set`.
+    if (LoadStoreOp* def =
+            dynamic_cast<LoadStoreOp*>(aliased_out->definition())) {
+      if (def != nullptr && def->opType() == LoadStoreOpType::SegmenterSet) {
+        continue;
+      }
+    }
+
+    // This is suboptimal in many uncommon cases. My head hurts when thinking
+    // about them, so go simple for now :)
+    //
+    // Legend:
+    //   M* = a meta op defining a **fusion output**
+    //   N/M = a non-meta op defining a **fusion output**
+    //
+    // Case 1:
+    //
+    //   N/M -> N/M
+    //      |
+    //      -->  M
+    //
+    // We should put a `segment_set` on the **edge** from N/M to M, so the two
+    // `N/M`s go to the same kernel.
+    //
+    // Case 2:
+    //
+    //   N/M -> M1 -> M2
+    //           |
+    //           --> N/M
+    //
+    // We should change it to
+    //
+    //   N/M -> M1 -> M2
+    //      |
+    //      --> M1' (non-output copy of M1) -> N/M
```

Comment:
This sorta depends on whether `N/M` is an IO tensor or not.

i.e. if N/M is just an intermediate, then the mutation doesn't really gain anything, whether outputting `M1` or `N/M`...

I'm just saying I agree with you here on what we could be missing out, and I'm happy with keeping it simple for now. :)

Reply:
That's right. My comment earlier in this section says `M*` and `N/M` are fusion outputs.